File: Robotics Kinematics and Dynamics Course/Chapter3/Lesson4/Chapter3_Lesson4.cpp
#include <Eigen/Dense>
#include <cmath>

using Matrix3 = Eigen::Matrix3d;
using Vector3 = Eigen::Vector3d;
using Quaternion = Eigen::Quaterniond;

Matrix3 axisAngleToR(const Vector3& omega_hat, double theta) {
    Vector3 w = omega_hat.normalized();
    Eigen::Matrix3d K;
    K << 0.0,   -w.z(),  w.y(),
          w.z(),  0.0,   -w.x(),
         -w.y(),  w.x(),  0.0;
    return Matrix3::Identity()
         + std::sin(theta) * K
         + (1.0 - std::cos(theta)) * K * K;
}

void RToAxisAngle(const Matrix3& R, Vector3& omega_hat, double& theta) {
    double trace = R.trace();
    double cos_theta = (trace - 1.0) / 2.0;
    cos_theta = std::max(-1.0, std::min(1.0, cos_theta));
    theta = std::acos(cos_theta);

    if (theta < 1e-8) {
        omega_hat = Vector3(1.0, 0.0, 0.0);
        theta = 0.0;
        return;
    }

    omega_hat = Vector3(
        R(2,1) - R(1,2),
        R(0,2) - R(2,0),
        R(1,0) - R(0,1)
    ) / (2.0 * std::sin(theta));
    omega_hat.normalize();
}

Quaternion RToQuat(const Matrix3& R) {
    Quaternion q;
    q = Quaternion(R);  // Eigen has built-in constructor
    q.normalize();
    return q;
}

Matrix3 quatToR(const Quaternion& q) {
    Quaternion qn = q.normalized();
    return qn.toRotationMatrix();
}

Matrix3 eulerZYXToR(double phi, double theta, double psi) {
    Matrix3 Rx, Ry, Rz;
    double c1 = std::cos(psi), s1 = std::sin(psi);
    double c2 = std::cos(theta), s2 = std::sin(theta);
    double c3 = std::cos(phi), s3 = std::sin(phi);

    Rz << c1, -s1, 0.0,
           s1,  c1, 0.0,
           0.0, 0.0, 1.0;
    Ry <<  c2, 0.0, s2,
            0.0, 1.0, 0.0,
           -s2, 0.0, c2;
    Rx << 1.0, 0.0, 0.0,
           0.0, c3, -s3,
           0.0, s3,  c3;

    return Rz * Ry * Rx;
}
      

File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter3/Lesson4/Chapter3_Lesson4.py
import numpy as np
from numpy.linalg import norm

def skew(omega):
    """Return skew-symmetric matrix [omega]_x."""
    wx, wy, wz = omega
    return np.array([[0.0, -wz,  wy],
                     [wz,  0.0, -wx],
                     [-wy, wx,  0.0]])

def axis_angle_to_R(omega_hat, theta):
    """Axis-angle to rotation matrix via Rodrigues."""
    omega_hat = np.asarray(omega_hat, dtype=float)
    omega_hat = omega_hat / norm(omega_hat)
    K = skew(omega_hat)
    return np.eye(3) + np.sin(theta) * K + (1.0 - np.cos(theta)) * (K @ K)

def R_to_axis_angle(R, eps=1e-8):
    """Rotation matrix to axis-angle."""
    R = np.asarray(R, dtype=float)
    # Ensure approximate orthogonality if needed
    trace = np.trace(R)
    # Clamp for numerical safety
    cos_theta = (trace - 1.0) / 2.0
    cos_theta = np.clip(cos_theta, -1.0, 1.0)
    theta = np.arccos(cos_theta)
    if theta < eps:
        return np.array([1.0, 0.0, 0.0]), 0.0
    if np.pi - theta < 1e-6:
        # Near pi: use diagonal entries
        diag = np.diag(R)
        omega = np.sqrt(np.maximum((diag + 1.0) / 2.0, 0.0))
        # Fix signs using off-diagonals
        if R[0, 1] < 0.0:
            omega[1] = -omega[1]
        if R[0, 2] < 0.0:
            omega[2] = -omega[2]
        omega_hat = omega / norm(omega)
        return omega_hat, theta
    omega_hat = np.array([
        R[2, 1] - R[1, 2],
        R[0, 2] - R[2, 0],
        R[1, 0] - R[0, 1]
    ]) / (2.0 * np.sin(theta))
    return omega_hat, theta

def quat_to_R(q):
    """Unit quaternion (w, x, y, z) to rotation matrix."""
    q = np.asarray(q, dtype=float)
    q = q / norm(q)
    w, x, y, z = q
    R = np.array([
        [1 - 2*(y*y + z*z), 2*(x*y - w*z),     2*(x*z + w*y)],
        [2*(x*y + w*z),     1 - 2*(x*x + z*z), 2*(y*z - w*x)],
        [2*(x*z - w*y),     2*(y*z + w*x),     1 - 2*(x*x + y*y)]
    ])
    return R

def R_to_quat(R):
    """Rotation matrix to quaternion (w, x, y, z)."""
    R = np.asarray(R, dtype=float)
    t = np.trace(R)
    if t > 0.0:
        s = np.sqrt(1.0 + t) * 2.0
        w = 0.25 * s
        x = (R[2, 1] - R[1, 2]) / s
        y = (R[0, 2] - R[2, 0]) / s
        z = (R[1, 0] - R[0, 1]) / s
    else:
        # Find the major diagonal term
        if R[0, 0] > R[1, 1] and R[0, 0] > R[2, 2]:
            s = np.sqrt(1.0 + R[0, 0] - R[1, 1] - R[2, 2]) * 2.0
            w = (R[2, 1] - R[1, 2]) / s
            x = 0.25 * s
            y = (R[0, 1] + R[1, 0]) / s
            z = (R[0, 2] + R[2, 0]) / s
        elif R[1, 1] > R[2, 2]:
            s = np.sqrt(1.0 + R[1, 1] - R[0, 0] - R[2, 2]) * 2.0
            w = (R[0, 2] - R[2, 0]) / s
            x = (R[0, 1] + R[1, 0]) / s
            y = 0.25 * s
            z = (R[1, 2] + R[2, 1]) / s
        else:
            s = np.sqrt(1.0 + R[2, 2] - R[0, 0] - R[1, 1]) * 2.0
            w = (R[1, 0] - R[0, 1]) / s
            x = (R[0, 2] + R[2, 0]) / s
            y = (R[1, 2] + R[2, 1]) / s
            z = 0.25 * s
    q = np.array([w, x, y, z])
    q /= norm(q)
    return q

def euler_zyx_to_R(phi, theta, psi):
    """Euler ZYX (roll, pitch, yaw) to rotation matrix."""
    c1, s1 = np.cos(psi), np.sin(psi)
    c2, s2 = np.cos(theta), np.sin(theta)
    c3, s3 = np.cos(phi), np.sin(phi)
    R = np.array([
        [c1*c2,            s3*s2*c1 - s1*c3,  s3*s1 + s2*c3*c1],
        [s1*c2,            s3*s1*s2 + c3*c1, -s3*c1 + s1*s2*c3],
        [-s2,              s3*c2,             c3*c2]
    ])
    return R

def R_to_euler_zyx(R, eps=1e-8):
    """Rotation matrix to Euler ZYX (roll, pitch, yaw)."""
    R = np.asarray(R, dtype=float)
    r31 = R[2, 0]
    if abs(r31) < 1.0 - eps:
        theta = np.arcsin(-r31)
        phi = np.arctan2(R[2, 1], R[2, 2])
        psi = np.arctan2(R[1, 0], R[0, 0])
    else:
        # Gimbal lock
        if r31 < 0:  # theta ~ +pi/2
            theta = np.pi / 2.0
            phi = 0.0
            psi = np.arctan2(-R[0, 1], R[1, 1])
        else:        # theta ~ -pi/2
            theta = -np.pi / 2.0
            phi = 0.0
            psi = np.arctan2(R[0, 1], R[1, 1])
    return phi, theta, psi

# Example: Using SciPy's Rotation class (if available)
try:
    from scipy.spatial.transform import Rotation as Rsc

    def quat_to_R_scipy(q):
        return Rsc.from_quat([q[1], q[2], q[3], q[0]]).as_matrix()

    def R_to_quat_scipy(R):
        r = Rsc.from_matrix(R)
        x, y, z, w = r.as_quat()
        return np.array([w, x, y, z])

except ImportError:
    pass
      
